crates/turbo-tasks/src/generics/index_map.rs

```diff
@@ -0,0 +1,71 @@
+use anyhow::Result;
+use indexmap::IndexMap;
+// This specific macro identifier is detected by turbo-tasks-build.
+use turbo_tasks_macros::generic_type as __turbo_tasks_internal_generic_type;
+
+use crate::{
+    self as turbo_tasks,
+    debug::{ValueDebug, ValueDebugFormat, ValueDebugString},
+    ValueDefault, Vc,
+};
+
+__turbo_tasks_internal_generic_type!(<K, V>, IndexMap<Vc<K>, Vc<V>>);
+
+#[turbo_tasks::function]
+async fn index_map_len(index_map: Vc<IndexMap<Vc<()>, Vc<()>>>) -> Result<Vc<usize>> {
+    let index_map = index_map.await?;
+    Ok(Vc::cell(index_map.len()))
+}
+
+#[turbo_tasks::function]
+async fn index_map_is_empty(index_map: Vc<IndexMap<Vc<()>, Vc<()>>>) -> Result<Vc<bool>> {
+    let index_map = index_map.await?;
+    Ok(Vc::cell(index_map.is_empty()))
+}
+
+impl<K, V> Vc<IndexMap<Vc<K>, Vc<V>>> {
+    /// See [`IndexMap::len`].
+    pub fn len(self) -> Vc<usize> {
+        index_map_len(Self::to_repr(self))
+    }
+
+    /// See [`IndexMap::is_empty`].
+    pub fn is_empty(self) -> Vc<bool> {
+        index_map_is_empty(Self::to_repr(self))
+    }
+}
+
+#[turbo_tasks::function]
+fn index_map_default() -> Vc<IndexMap<Vc<()>, Vc<()>>> {
+    Vc::cell(Default::default())
+}
+
+impl<K, V> ValueDefault for IndexMap<Vc<K>, Vc<V>> {
+    fn value_default() -> Vc<Self> {
+        // Safety: `index_map_default` creates a None variant, which is a valid
+        // representation of any index set of `Vc`.
+        unsafe { Vc::<Self>::from_repr(index_map_default()) }
+    }
+}
+
+#[turbo_tasks::function]
+async fn index_map_dbg_depth(
+    index_map: Vc<IndexMap<Vc<()>, Vc<()>>>,
+    depth: usize,
+) -> Result<Vc<ValueDebugString>> {
+    Ok(index_map
+        .await?
+        .value_debug_format(depth)
+        .try_to_value_debug_string()
+        .await?)
+}
```

Comment:
We only cast them from the Rust compiler's point of view, but the `Vc`s themselves still hold their actual type information. Hence, this will eventually hit turbo_tasks' runtime implementation resolving, which calls the correct `Vc` debug implementation.

The following implementations get hit in order:
1. `impl ValueDebugFormat for IndexMap<(), ()>`
2. `impl ValueDebugFormat for Vc<()>`
3. `impl ValueDebug for Box<dyn ValueDebug>`
4. `turbo_tasks::trait_call(...)` <- this reads the type from the cell directly and uses that to dispatch the method.